modules/core/04-channel/keeper/upgrade.go

```diff
@@ -597,6 +597,94 @@ func (k Keeper) startFlushUpgradeHandshake(
 	return nil
 }
 
+// ChanUpgradeOpen is called by a module to complete the channel upgrade handshake and move the channel back to an OPEN state.
+// This method should only be called after both channels have flushed any in-flight packets.
+func (k Keeper) ChanUpgradeOpen(
+	ctx sdk.Context,
+	portID,
+	channelID string,
+	counterpartyChannelState types.State,
+	proofChannel []byte,
+	proofHeight clienttypes.Height,
+) error {
+	channel, found := k.GetChannel(ctx, portID, channelID)
+	if !found {
+		return errorsmod.Wrapf(types.ErrChannelNotFound, "port ID (%s) channel ID (%s)", portID, channelID)
+	}
+
+	if k.hasInflightPackets(ctx, portID, channelID) {
+		return errorsmod.Wrapf(types.ErrPendingInflightPackets, "port ID (%s) channel ID (%s)", portID, channelID)
+	}
+
+	if !collections.Contains(channel.State, []types.State{types.TRYUPGRADE, types.ACKUPGRADE}) {
+		return errorsmod.Wrapf(types.ErrInvalidChannelState, "expected one of [%s, %s], got %s", types.TRYUPGRADE, types.ACKUPGRADE, channel.State)
+	}
+
+	if channel.FlushStatus != types.FLUSHCOMPLETE {
+		return errorsmod.Wrapf(types.ErrInvalidFlushStatus, "expected %s, got %s", types.FLUSHCOMPLETE, channel.FlushStatus)
+	}
+
+	connection, err := k.GetConnection(ctx, channel.ConnectionHops[0])
+	if err != nil {
+		return errorsmod.Wrap(err, "failed to retrieve connection using the channel connection hops")
+	}
+
+	if connection.GetState() != int32(connectiontypes.OPEN) {
+		return errorsmod.Wrapf(connectiontypes.ErrInvalidConnectionState, "connection state is not OPEN (got %s)", connectiontypes.State(connection.GetState()).String())
+	}
+
+	var counterpartyChannel types.Channel
+	counterpartyHops := []string{connection.GetCounterparty().GetConnectionID()}
+	switch counterpartyChannelState {
+	case types.OPEN:
+		upgrade, found := k.GetUpgrade(ctx, portID, channelID)
+		if !found {
+			return errorsmod.Wrapf(types.ErrUpgradeNotFound, "port ID (%s) channel ID (%s)", portID, channelID)
+		}
+		counterpartyChannel = types.Channel{
```

Comment:
nit for white spacing

```suggestion
		}
		
		counterpartyChannel = types.Channel{
```